java-checks/src/main/java/org/sonar/java/checks/BoxedBooleanExpressionsCheck.java

```diff
@@ -127,17 +135,61 @@ private boolean isSafeBooleanExpression(ExpressionTree tree) {
   private static boolean isFirstUsageANullCheck(ExpressionTree boxedBoolean) {
     if (boxedBoolean.is(Kind.IDENTIFIER)) {
       IdentifierTree identifier = (IdentifierTree) boxedBoolean;
+      // Usages are not guaranteed to be ordered
       List<IdentifierTree> usages = identifier.symbol().usages();
-      if (!usages.isEmpty()) {
-        Tree parent = usages.get(0).parent();
-        if (parent.is(Kind.NOT_EQUAL_TO, Kind.EQUAL_TO) && isNullCheck((ExpressionTree) parent)) {
-          return true;
-        }
+      Tree firstUsage = usages.get(0).parent();
+      // Test if the first usage in our list is a null check
+      if (firstUsage.is(Kind.EQUAL_TO, Kind.NOT_EQUAL_TO) && isNullCheck((ExpressionTree) firstUsage)) {
+        return true;
+      }
+      // Return false if the only usage is not a null check
+      if (usages.size() == 1) {
+        return false;
+      }
+      // Fetch the first null check in the usages list
+      Optional<ExpressionTree> firstNullCheck = getFirstNullCheck(identifier);
+      if (!firstNullCheck.isPresent()) {
+        return false;
       }
+      // Test if the first null check and the first usage are part of the same higher if structure
+      Optional<IfStatementTree> ifStatementWithNullCheck = getParentConditionalBranch(firstNullCheck.get());
+      Optional<IfStatementTree> ifStatementWithFirstUsage = getParentConditionalBranch((ExpressionTree) firstUsage);
+      return ifStatementWithNullCheck.equals(ifStatementWithFirstUsage);
     }
     return false;
   }
 
+  private static Optional<ExpressionTree> getFirstNullCheck(IdentifierTree identifier) {
+    if (firstNullCheckCache.containsKey(identifier)) {
+      return firstNullCheckCache.get(identifier);
+    }
+    List<IdentifierTree> usages = identifier.symbol().usages();
+    Optional<ExpressionTree> firstNullCheck = getFirstNullCheck(usages);
+    firstNullCheckCache.put(identifier, firstNullCheck);
+    return firstNullCheck;
+  }
+
+  private static Optional<ExpressionTree> getFirstNullCheck(List<IdentifierTree> usages) {
+    return usages.stream()
+      .map(IdentifierTree::parent)
+      .filter(tree -> tree.is(Kind.EQUAL_TO, Kind.NOT_EQUAL_TO) && isNullCheck((ExpressionTree) tree))
+      .map(ExpressionTree.class::cast)
+      .findFirst();
+  }
+
+  private static Optional<IfStatementTree> getParentConditionalBranch(ExpressionTree tree) {
+    if (ifStatementCache.containsKey(tree)) {
+      return ifStatementCache.get(tree);
+    }
+    Tree parent = tree;
+    while (parent != null && !parent.is(Kind.IF_STATEMENT)) {
+      parent = parent.parent();
```

Comment:
the point of caching here is that you might have already calculated the 'IfStatement parent' for some parent of this tree hierarchy. So you don't need to go through the whole loop of parents, just find the first parent. whose closest IfStatement we already know.

Reply:
The cache is now stores all the parents leading up to an `IfStatementTree` are cached at the end of the method.
We also add parents that lead nowhere to make better use of the cache.